### PR TITLE
Switch to slack_sdk and add support to ignore users and send files

### PR DESCRIPTION
**Pull request by BrianGallew**
This is huge, I know.  Sorry.

Summary:

* pep8 compliance of changed files
* housekeeping (adding doc-strings and remove unused imports)
* Rewrite slack backend to use slack_sdk 
* Drop support from python 2.7, moving support up to 3.6+
* Ability to ignore users

### CODE REVIEW

**Comment by BrianGallew** [comment]
Addresses #426

File: will/backends/io_adapters/base.py [hunk 1]
```diff
@@ -40,6 +40,11 @@ def normalize_incoming_event(self, event):
 
     def handle_incoming_event(self, event):
         try:
+            user_id = event.get('user', 'wHATEver')
```
Comment (rsalmond) [inline will/backends/io_adapters/base.py:43]: Probably just being thick but the need for this constant isn't obvious to me, what's it for?

Comment (BrianGallew) [reply will/backends/io_adapters/base.py:43]: That's me not trusting that every possible incoming event will actually have a 'user' element.

**Comment by BrianGallew** [comment]
Forgot I was supposed to tag @Ashex with this.

**Comment by BrianGallew** [comment]
I suppose I should point out: this code is what we've been running in production for some time now (most of it for more than 3 months).  The ability to ignore users is newer, though: only 3 weeks or so.  We had a couple of bots get into a loop, and this was an easy fix.

File: will/backends/io_adapters/base.py (hunk 1 above)
Comment (rsalmond) [reply will/backends/io_adapters/base.py:43]: Er, sorry should have been more specific. Does setting it to `wHATEver` when the user element is missing do something different than allowing it to be `None` ?

File: will/plugin.py [hunk 2]
```diff
@@ -51,23 +57,23 @@ def get_backend(self, message, service=None):
...
-        logging.info("self.say")
-        logging.info(content)
+        'Publish an event to be shipped to one of the IO backends.'
         if channel:
             room = channel
         elif room:
             channel = room
 
-        if not "channel" in kwargs and channel:
+        if "channel" not in kwargs and channel:
```
Comment (rsalmond) [inline will/plugin.py:76]: :heart:

Comment (BrianGallew) [reply will/plugin.py:76]: Yeah, there's a significant amount of edits that boil down to "pep-8 and comments".  I also added a non-trivial amount of type annotations.

File: will/backends/io_adapters/slack.py [hunk 3]
```diff
@@ -457,85 +619,75 @@ def _update_channels(self):
...
-                time.sleep(2)
-            except (KeyboardInterrupt, SystemExit):
-                break
-            except:
-                logging.critical("Error in watching slack RTM: \n%s" % traceback.format_exc())
-                break
+        "This is our main loop."
+        # The decorators don't work on unbound methods.  Sigh.
+        # These are all events that should spark an update of our inventory
+        # RTMClient.run_on(event="hello")(self._update_backend_metadata)
```
Comment (rsalmond) [inline will/backends/io_adapters/slack.py:674]: Can we eliminate these lines or is the intention to execute these updates on other event types in the future?

File: will/backends/io_adapters/base.py (hunk 1 above)
Comment (BrianGallew) [reply will/backends/io_adapters/base.py:43]: ... nope.  Sigh.

File: will/backends/io_adapters/slack.py (hunk 3 above)
Comment (BrianGallew) [reply will/backends/io_adapters/slack.py:674]: Good question.  Some of these events we absolutely don't care about.  Others, we really should do an incremental update rather than a full update.  Yet at the same time, we probably want to do a full update periodically just to be sure we didn't miss anything.  The docs say that the "hello" event should be the first one you get, and that you should load data on that.  I did not find that to be the case.  So I left in these commented events as hints to myself, but really, they're probably not useful to anyone but myself.

File: will/backends/io_adapters/base.py (hunk 1 above)
Comment (rsalmond) [reply will/backends/io_adapters/base.py:43]: Gotcha, just wanted to make sure I wasn't missing something. Thanks.

File: will/__init__.py [hunk 4]
```diff
@@ -1 +1 @@
-VERSION = "2.1.3"
+VERSION = "4.0.0"
```
Comment (Ashex) [inline will/__init__.py:1]: I'd say we're due for a new release but not this major of one. Lets be conservative with 2.2.0 to mark a major rewrite of the slack backend. @skoczen then we can push a full release to pypi.

File: will/backends/encryption/aes.py [hunk 5]
```diff
@@ -1,65 +1,67 @@
...
+        except Exception:
+            logging.exception("Error preparing message for the wire: \n%s", raw)
             return None
 
-    @classmethod
-    def decrypt_from_b64(cls, raw_enc):
+    @staticmethod
+    def decrypt_from_b64(enc):
+        'decrypt b64-encoded data'
         try:
```
Comment (Ashex) [inline will/backends/encryption/aes.py:51]: can you explain the change to the `decrypt_from_b64` method?

File: will/backends/io_adapters/base.py (hunk 1 above)
Comment (Ashex) [reply will/backends/io_adapters/base.py:43]: I think `user` is missing when it's a bot or webhook. As the intent here is to just do a list lookup can we go with an empty string instead?

File: will/plugin.py [hunk 6]
```diff
@@ -51,23 +57,23 @@ def get_backend(self, message, service=None):
...
 
     def get_message(self, message_passed):
+        'Try to find a message'
         if not message_passed and hasattr(self, "message"):
             return self.message
         return message_passed
 
     def say(self, content, message=None, room=None, channel=None, service=None, package_for_scheduling=False, **kwargs):
-        logging.info("self.say")
-        logging.info(content)
```
Comment (Ashex) [inline will/plugin.py:64]: Nothing to say aside from this is a welcome change to not dump the content on _every single message_.

File: will/plugin.py [hunk 7]
```diff
@@ -83,54 +89,82 @@ def say(self, content, message=None, room=None, channel=None, service=None, pack
...
             if package_for_scheduling:
                 return "message.outgoing.%s" % backend, e
-            else:
-                logging.info("putting in queue: %s" % content)
-                self.publish("message.outgoing.%s" % backend, e)
+            logging.info("putting in queue: %s", content)
+            self.publish("message.outgoing.%s" % backend, e)
+        return None
+
+    def send_file(self, message, content, filename, text=None, filetype='text',
```
Comment (Ashex) [inline will/plugin.py:96]: Let's split file upload to a separate PR as I know at least a few of us have written our own implementation (for example #372 )

File: will/plugins/help/programmer_help.py [hunk 8]
```diff
@@ -1,15 +1,16 @@
...
         all_regexes = self.load("all_listener_regexes")
-        help_text = "Here's everything I know how to listen to:"
-        for r in all_regexes:
-            help_text += "\n%s" % r
-
-        self.say(help_text, message=message)
+        self.say("Here's everything I know how to listen to:", message, start_thread=True)
+        for r in range(0, len(all_regexes), MAX_LINES):
+            text = "\n".join(all_regexes[r:r+MAX_LINES])
+            self.say(f'```{text}```', message, start_thread=True)
```
Comment (Ashex) [inline will/plugins/help/programmer_help.py:16]: ♥

File: will/requirements/slack.txt [hunk 9]
```diff
@@ -1,4 +1,4 @@
 -r base.txt
-slackclient>=1.2.1,<1.3.0
-markdownify==0.4.1
-
+slack-sdk==3.0.0
+markdownify==0.5.3
+aiohttp
```
Comment (Ashex) [inline will/requirements/slack.txt:4]: I don't see the added use of aiohttp in the PR so let's remove it from the requirements file.

**Review by Ashex — CHANGES_REQUESTED** [review]
MANY welcome changes here. We just need to get a little clarity on some changes and remove a couple to keep this PR scope "narrow".

File: will/backends/encryption/aes.py (hunk 5 above)
Comment (BrianGallew) [reply will/backends/encryption/aes.py:51]: The original declaration didn't actually reference the class in any way.

**Comment by BrianGallew** [comment]
I've pushed an update that should address all of the raised concerns.

File: will/plugin.py [hunk 10]
```diff
@@ -6,14 +7,15 @@
 from will import settings
 from will.abstractions import Event, Message
 # Backwards compatability with 1.x, eventually to be deprecated.
-from will.backends.io_adapters.hipchat import HipChatRosterMixin, HipChatRoomMixin
 from will.mixins import NaturalTimeMixin, ScheduleMixin, StorageMixin, SettingsMixin, \
     EmailMixin, PubSubMixin
-from will.utils import html_to_text
 
+FILENAME_CLEANER = re.compile(r'[^-_0-9a-zA-Z]+')
```
Comment (Ashex) [inline will/plugin.py:13]: Once last bit, this is not referenced so we should remove it from this change.

**Comment by Ashex** [comment]
Another note for reference, the introduction of f-strings will force us to only support python 3.6+ therefore satisfying #417

File: will/backends/io_adapters/slack.py [hunk 11]
```diff
@@ -142,99 +188,123 @@ def normalize_incoming_event(self, event):
...
-            if hasattr(event, "source_message") and event.source_message and "channel" not in kwargs:
-                self.send_message(event)
-            else:
-                # Came from webhook/etc
-                # TODO: finish this.
-                target_channel = kwargs.get("room", kwargs.get("channel", None))
-                if target_channel:
-                    event.channel = self.get_channel_from_name(target_channel)
-                    if event.channel:
+    def send_file(self, event):
```
Comment (Ashex) [inline will/backends/io_adapters/slack.py:204]: Just getting to a closer look at the slack backend. Let's also remove `send_file` here. We definitely need file upload support so if you're able to submit this as a separate PR we can review separetely and get it merged in before we cut a 2.2.0 release.

File: will/backends/io_adapters/slack.py [hunk 12]
```diff
@@ -457,85 +619,72 @@ def _update_channels(self):
...
+        if client:
+            for page in client.users_list(limit=self.PAGE_LIMIT):
+                for member in page["members"]:
+                    if member["deleted"]:
+                        continue
+                    member_id = member["id"]
+                    user_timezone = member.get("tz")
+                    people[member_id] = Person(
+                        id=member_id,
+                        mention_handle=member.get("mention_handle", "SECRETAGENTMAN"),
```
Comment (Ashex) [inline will/backends/io_adapters/slack.py:634]: I love having this as the fallback value but we should probably use an empty string here and at line 637.

File: will/backends/io_adapters/slack.py [hunk 13]
```diff
@@ -457,85 +619,72 @@ def _update_channels(self):
...
+        if name is not None:
+            self.__class__.handle = name
+        Thread(
+            target=self._update_people, args=(event["web_client"],), daemon=True
+        ).start()
+        Thread(
+            target=self._update_channels, args=(event["web_client"],), daemon=True
+        ).start()
 
     def _watch_slack_rtm(self):
```
Comment (Ashex) [inline will/backends/io_adapters/slack.py:670]: It looks like this should resolve any issues where an unhandled exception aborts the loop.

**Review by Ashex — CHANGES_REQUESTED** [review]
Reviewed the slack backend specifically (that I missed on my first run through as github didn't render the diff). This is a monumental rewrite and it looks great. I don't want to rush getting this merged in so I'll leave the rest of the contributors to review and provide feedback through the weekend. 

I'll spin up a test instance with this branch on monday and run will through some basic tests to validate.

File: will/backends/io_adapters/slack.py (hunk 11 above)
Comment (BrianGallew) [reply will/backends/io_adapters/slack.py:204]: I think you need to make up your mind here.  You wanted me to pull send_file() out because there are (possibly) several alternative implementations and an existing PR.  That's fine, but *I* won't be submitting a separate PR for this.  Splitting things into multiple PRs just for the sake of multiple PRs adds pointless burden to me.  I'll pull this out, and then you can look at code submitted by someone else.

Comment (Ashex) [reply will/backends/io_adapters/slack.py:204]: @pastorhudson @rsalmond what are your thoughts here? I'm of the mind that we need to keep this PR limited to rewriting the backend to support current methods and the file upload is a great bonus _however_ I'm wary of bringing in changes that will break other implementations.

Comment (pastorhudson) [reply will/backends/io_adapters/slack.py:204]: I get what you’re saying, but I think if we want file support and this works then merge it in. At this point this project is almost dead because of stagnation and lack of merging pull requests. For the health of the project I think we’d be better to merge some stuff even if it’s not quite perfect just to keep the ball rolling. We can always fix bugs if we discover them as long as things are getting regularly merged.

Comment (BrianGallew) [reply will/backends/io_adapters/slack.py]: Removed

File: will/backends/io_adapters/slack.py (hunk 12 above)
Comment (BrianGallew) [reply will/backends/io_adapters/slack.py]: Removed

**Comment by BrianGallew** [comment]
The send_file code is all gone at this point.

**Comment by skoczen** [comment]
My two cents: 

There's precedent to backends having exclusive features, if they're not something that can be abstracted generally, or they're not common, etc.

Merging with files and then building up an abstraction from there would seem sane to me.

That said, keeping things moving does seem to be the highest priority.


@BrianGallew Thanks for sending this.  As you know, it's a community effort: folks here (including me) are fitting this in between their jobs, their families, relationships, doctor's appointments, etc.   It takes time to build consensus, and asynchronous, community-based PRs take extra time.   Your patience is appreciated, and it makes things _much_ easier.   Also, it's hard to tell tone on the internet.  Please give a little extra kindness and patience to the process, like I'm sure you'd appreciate if you were on the other side!

@Ashex @pastorhudson @rsalmond  - do we need any changes to contributor/owner status for the repo or pypi to get this out? I've got for sure solid internet for the next couple of days, and really happy to get it done.

**Comment by pastorhudson** [comment]
I'm happy to do whatever you guys need

**Comment by BrianGallew** [comment]
I'm not angry.  Like the rest of you, I'm busy with my own life.  I'm quite happy to contribute to will.  I'm simply uninterested in breaking things out into multiple commits/PRs.  If the send_file code I submitted was unacceptable for any reason, then there's no reason for me to re-submit it.  As Ashex has pointed out, there's an existing PR covering that, and potentially other implementations out there to choose from.

**Comment by Ashex** [comment]
@BrianGallew Based off everyones feedback can you add the send_file implementation back? I truly appreciate the work you've done and am grateful for this pull request, my aim was to not disrupt the work others have done but at the end of the day those are custom changes and the PR I referenced has been sitting for some time. 

@skoczen I'll need pypi access so I can cut a 2.2.0 release with this PR and everything missing since 2.1.3 was released.

**Comment by BrianGallew** [comment]
send_file is back in and I *think* I kept all of the other requested changes.

**Comment by Ashex** [comment]
Can you also add it back in `will/plugin.py`? I believe this is the function:

https://github.com/skoczen/will/pull/428/files/40686b2f66e21aeec17275fd2c7bc4e767170e0b#diff-91352b3c9c5fadae5d1a560fbecb3204f07366815f315540f37b8b8db7da9461R96

**Comment by BrianGallew** [comment]
Sorry, my weak git-fu is now revealed.  I had actually reverted *all* my changes, including the ones I had wanted to retain.  They're in there now, and according to grep send_file is now declared in both slack.py and plugin.py.

**Comment by pastorhudson** [comment]
> Sorry, my weak git-fu is now revealed. I had actually reverted _all_ my changes, including the ones I had wanted to retain. They're in there now, and according to grep send_file is now declared in both slack.py and plugin.py.

No worries. It happens to us all!

**Comment by Ashex** [comment]
No worries at all! I'll commence testing tomorrow :) I'll write a plugin for executing a variety of actions and share the gist with the approval.

**Comment by skoczen** [comment]
@Ashex Yep, you got it on Pypi - what's your pypi username?

**Comment by Ashex** [comment]
@skoczen My pypi username is [Ashex](https://pypi.org/user/Ashex/)

**Review by pastorhudson — APPROVED** [review]

**Comment by skoczen** [comment]
@Ashex you've been added to Pypi, and should be able to release as soon as you accept the invite!

**Review by Ashex — CHANGES_REQUESTED** [review]
I ran this through some tests, the only thing that needs to be fixed is adding this back to `will/plugin.py` as otherwise send_file fails.

`FILENAME_CLEANER = re.compile(r'[^-_0-9a-zA-Z]+')`

And putting aiohttp back in as it turns out it is needed since we're using the async api. The line should be `aiohttp>=3.7.3,<4` in order to pin the version.


Here's the plugin I wrote to do some quick tests, I wanted to ask if send_file supports DM as I was not able to get that to function.

https://gist.github.com/Ashex/cb66194c429fb8d4d2f19a15fcc75b64

There's a couple minor things I spotted but I'll correct that myself through another PR.

**Comment by Ashex** [comment]
Great, can you also add aiohttp back to the requirements file? I'll get this merged in once that's done.

**Comment by Ashex** [comment]
Regarding DM support, it does work when interacting with the bot directly. There's just some extra magic required when you want to send a user an unsolicited DM with a file, I did sort it out elsewhere so expect a PR in the near future with the enhancement.

**Comment by BrianGallew** [comment]
aoihttp change reverted

**Review by pastorhudson — APPROVED** [review]

**Review by Ashex — APPROVED** [review]
Thank you again for your patience and all the work you put in to rewrite this, I can definitely say you've brought life back to the project.

I'll now approve and get this merged. We have some work left to do before we cut a new release with this but it will be coming soon.

**Comment by pastorhudson** [comment]
I totally agree with @Ashex thank you for the work you put into this @BrianGallew! I really appreciate it. I’m excited to see this project moving again! 🥳

**Comment by BrianGallew** [comment]
:victory:
